modules/standard/Socket.md

```diff
@@ -0,0 +1,82 @@
+## Socket Module API
+
+To those familiar with the Unix socket API, the method names will feel familiar, though their usage will be somewhat simpler than the raw Unix socket API.
+
+## Creating a Listener
+
+`socket.listen` function will create a `TCPListener` Instance which will take in following parameters
+
+```chapel
+socket.listen([addr:IPAddr,]port:int,backlog=5,reuseaddr=true)
+```
+
+The `TCPListener` instance will have methods on it to call
+- `accept`,
+- `close`,
+- `addr` and
+- `setSocketOpt` to set various socket options.
+
+`accept` method will return a `TCPConn` instance and associated `addr` of the connection.
+
+## Connecting to remote sockets
+
+`socket.connect` function will create `TCPConn` instance which will take in the following parameters
+
+```chapel
+const client = socket.connect(addr:IPAddr,port:int)
+```
+
+A higher level construct for socket will take in addr as `string` which will we resolved and converted to `IPAddr` internally.
+
+```chapel
+const client = socket.connect(addr:string,port:int)
+```
+
+The `TCPConn` instance will have methods on it to call
+- `read` ,
+- `write` ,
+- `close`,
+- `closeWrite`,
+- `closeRead`,
+- `setKeepAlive`,
+- `addr` ,
+- `setReadBuffer`
+- and `setWriteBuffer`
+
+The additional support with `TCPConn` and `TCPListener` will be to enable or disable two of the prominent optimization algorithm for TCP Protocol that are:
+
+- `nagle`
+- `quickack`
+
+The support will be in form of separate utility function attached with socket module that can set/unset the optimization based on `boolean` argument
+
+## UDP Support
+
+Support for UDP protocol will be through a separate constructor for `UDPSocket`
+
+A generic `bind` function will be used which will take in any of the socket three object and bind it to provided `host` , `port`.
```

Comment:
Wouldn't this bind function be defined in a base class ? How will the bind function differentiate between ipv4 and ipv6? Please write more about the base class and what other methods including bind will be present in it.

Reply:
- `bind` will be defined at the module level
- it can differentiate btw ipv4 and ipv6 based on the type of host user provides to the function call

Reply:
I don't really remember what `bind` does in the system call, but can you describe it here for anybody else who has forgotten? Is it only used for UDP because it is "connectionless" ? If it's just setting the address we are listening on, why wouldn't we do that when constructing the listener? Why do we have UDPSocket as a new type here - do we need a separate UDPSocket/TCPSocket at all? Would it be enough to have UDPListener / TCPListener and UDPConnection / TCPConnection?

Reply:
`bind` is responsible for binding any socket to a specified address and port. Here the socket can be anything either a UDP or a TCP Socket. `listen` under the hood will can in `bind` to bind `TCPListener` to address and port. With either `TCPConn` or `UDPSocket` that isn't a requirement as the system assigns any free port and localhost address to them by default.

While `TCPConn` is in a connected state when they are created `UDPSocket` are connectionless they are just waiting for the user to use them for `sending` and `receiving` purpose only. But for `sending` and receiving purpose user might require a socket to have a _known_ address that is when `bind` will be useful for `UDPSocket`.